### PR TITLE
kmeans: fix for handling NaN values (MADLIB-364)

### PR DESCRIPTION
**Pull request by agorajek**
Documentation update is included in Florian's latest doc commit.

### CODE REVIEW

File: src/ports/postgres/modules/kmeans/kmeans.py_in [hunk 1]
```diff
@@ -787,14 +787,14 @@ def kmeans( madlib_schema
...
     __run_quietly( sql);
     # Create the Temporary Point table
-    # Remove any NULL values
+    # Remove any NULL, NaN, and any non-finite values
     sql = '''
         INSERT INTO TempPoints0 
         SELECT pid, coords, 0, null
         FROM ''' + src_view + '''
-        WHERE ''' + madlib_schema + '.svec_dot(coords,coords) IS NOT NULL';
+        WHERE abs(''' + madlib_schema + '.svec_elsum(coords)) >= 0';
```
Comment (cwelton) [inline src/ports/postgres/modules/kmeans/kmeans.py_in]: -- What about vectors containing negative values?  I don't think el_sum >= 0 is the right condition here. --

Which is, of course, why you have the abs() there.  Never mind earlier comment.

Comment (cwelton) [inline src/ports/postgres/modules/kmeans/kmeans.py_in]: However I get this result:

```
 select 'NaN'::float8 >= 0;
 column
 ----------
  t
 (1 row)
```

Which means that I'm not convinced that this test will exclude the NaN's correctly.

File: src/ports/postgres/modules/kmeans/kmeans.py_in [hunk 2]
```diff
@@ -810,7 +810,7 @@ def kmeans( madlib_schema
         plpy.error( "input points must have the same dimensions\n")            
     info( ' * points: ' + str(orig_point_count) + ' (' \
             + str(rv[0]['min_dim']) + ' dimensions), kept ' + str(point_count) + \
-            ' after removing NULLs');     
+            ' after removing non-finite values');
```
Comment (cwelton) [inline src/ports/postgres/modules/kmeans/kmeans.py_in:815]: 'Infinity' is a non-finite value that would pass your >= 0 check as well.

File: src/ports/postgres/modules/kmeans/kmeans.py_in (hunk 1 above)
Comment (agorajek) [inline src/ports/postgres/modules/kmeans/kmeans.py_in]: NaN is greater than zero ?:

``` sql
test=# select 'NaN'::float8 > 0;
 ?column? 
----------
 t

test=# select 'NaN'::float8 = 0;
 ?column? 
----------
 f

test=# select 'NaN'::float8 < 0;
 ?column? 
----------
 f
```

Isn't this a bug?

Comment (fschopp) [inline src/ports/postgres/modules/kmeans/kmeans.py_in]: Yikes! GP is not IEEE 754 compliant.

**Comment by agorajek** [comment]
This commit should fix the problems. Also added a test.

**Comment by cwelton** [comment]
As stands I'd like to hold off on this until after the release.

**Comment by agorajek** [comment]
Why?

**Comment by cwelton** [comment]
Because it has to release tomorrow and the solution provided here does not look complete.

**Comment by cwelton** [comment]
Plus, I'm fine release noting this issue.

File: src/ports/postgres/modules/kmeans/kmeans.py_in (hunk 1 above)
Comment (fschopp) [inline src/ports/postgres/modules/kmeans/kmeans.py_in]: None of this is an efficient way of checking finiteness of all elements. It's also quite convoluted. That said, the correct (but still ugly) way would be something like

```
WHERE abs(coalesce(''' + madlib_schema + '.svec_elsum(coords), 'Infinity'::FLOAT8)) < 'Infinity'::FLOAT8
```

The above does not catch Infinity.

**Comment by agorajek** [comment]
1) Fine that it's a bug, but we would need a workaround in the interim.
2) Really?

``` sql
test=# select madlib.svec_elsum('{1,1}:{1,NaN}'::madlib.svec);
 svec_elsum 
------------
        NaN
```

3) Nice! It looks like it will work for NaNs as well.

**Comment by agorajek** [comment]
Fixed again.

File: src/ports/postgres/modules/kmeans/kmeans.py_in (hunk 1 above)
Comment (cwelton) [inline src/ports/postgres/modules/kmeans/kmeans.py_in]: Regarding IEEE 754 compliance, from the postgres documentation:

> Note: In most implementations of the “not-a-number” concept, NaN is not considered equal to any other numeric value (including NaN). In order to allow numeric values to be sorted and used in tree-based indexes, PostgreSQL treats NaN values as equal, and greater than all non-NaN values.

**Comment by cwelton** [comment]
Pulling after review

**Pull request merged**